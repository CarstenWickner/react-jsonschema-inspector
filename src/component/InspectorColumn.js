import PropTypes from "prop-types";
import React from "react";
import classNames from "classnames";

import InspectorItem from "./InspectorItem";
import { getColumnDataPropTypeShape } from "./renderDataUtils";

const InspectorColumn = (props) => {
    const {
        items, selectedItem, filteredItems, trailingSelection, onSelect, renderItemContent
    } = props;
    return (
        <div
            className={classNames({
                "jsonschema-inspector-column": true,
                "with-selection": selectedItem,
                "trailing-selection": trailingSelection
            })}
            onClick={onSelect}
            role="presentation"
            tabIndex={-1}
        >
            {Object.keys(items).sort().map(name => (
                <InspectorItem
                    key={name}
                    name={name}
                    schemaGroup={items[name]}
                    selected={name === selectedItem}
                    matchesFilter={filteredItems ? filteredItems.includes(name) : undefined}
                    onSelect={event => onSelect(event, name)}
                    renderContent={renderItemContent}
                />
            ))}
        </div>
    );
};

const columnDataPropTypeShape = getColumnDataPropTypeShape(false);
InspectorColumn.propTypes = {
    items: columnDataPropTypeShape.items.isRequired,
    selectedItem: PropTypes.string,
    filteredItems: PropTypes.arrayOf(PropTypes.string),
    trailingSelection: columnDataPropTypeShape.trailingSelection,
    onSelect: columnDataPropTypeShape.onSelect.isRequired,
    renderItemContent: PropTypes.func // func({ string: name, boolean: hasNestedItems, boolean: selected, JsonSchema: schema })
};
InspectorColumn.defaultProps = {
    selectedItem: null,
    filteredItems: null,
    trailingSelection: false,
    renderItemContent: null
};

export default InspectorColumn;
