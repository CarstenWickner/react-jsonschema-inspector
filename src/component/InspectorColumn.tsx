import * as React from "react";
import classNames from "classnames";

import { InspectorItem } from "./InspectorItem";
import { InspectorProps, RenderItemsColumn } from "./InspectorTypes";

export const InspectorColumn: React.FunctionComponent<{
    items: RenderItemsColumn["items"];
    onSelect: RenderItemsColumn["onSelect"];
    selectedItem?: RenderItemsColumn["selectedItem"];
    filteredItems?: RenderItemsColumn["filteredItems"];
    trailingSelection?: RenderItemsColumn["trailingSelection"];
    renderItemContent?: InspectorProps["renderItemContent"];
}> = ({ items, selectedItem, filteredItems, trailingSelection, onSelect, renderItemContent }): React.ReactElement => (
    <div
        className={classNames({
            "jsonschema-inspector-column": true,
            "with-selection": selectedItem,
            "trailing-selection": trailingSelection
        })}
        onClick={onSelect}
        role="presentation"
        tabIndex={-1}
    >
        {Object.keys(items)
            .sort()
            .map((name) => (
                <InspectorItem
                    key={name}
                    name={name}
                    schemaGroup={items[name]}
                    selected={name === selectedItem}
                    matchesFilter={filteredItems ? filteredItems.includes(name) : undefined}
                    onSelect={(event): void => onSelect(event, name)}
                    renderContent={renderItemContent}
                />
            ))}
    </div>
);
