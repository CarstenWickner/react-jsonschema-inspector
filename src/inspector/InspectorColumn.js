import PropTypes from 'prop-types';
import React, { PureComponent } from 'react';
import classNames from 'classnames';

import JsonSchemaPropType from './JsonSchemaPropType';
import InspectorItem from './InspectorItem';

class InspectorColumn extends PureComponent {
    renderItem(name) {
        const { items, selectedItem, refTargets, onSelect, renderItemContent } = this.props;
        const item = items[name];
        return (
            <InspectorItem
                key={name}
                name={name}
                schema={item}
                selected={name === selectedItem}
                refTargets={refTargets}
                onSelect={event => onSelect(event, name)}
                renderContent={renderItemContent}
            />
        );
    }

    render() {
        const { items, selectedItem, trailingSelection, onSelect } = this.props;
        const columnClassName = classNames({
            'jsonschema-inspector-column': true,
            'with-selection': selectedItem,
            'trailing-selection': trailingSelection
        });
        return (
            <div className={columnClassName} onClick={onSelect}>
                {Object.keys(items).sort().map(this.renderItem.bind(this))}
            </div>
        );
    }
}

InspectorColumn.propTypes = {
    items: PropTypes.objectOf(PropTypes.oneOfType([PropTypes.bool, JsonSchemaPropType])).isRequired,
    selectedItem: PropTypes.string,
    trailingSelection: PropTypes.bool,
    refTargets: PropTypes.objectOf(JsonSchemaPropType),
    onSelect: PropTypes.func.isRequired, // func(SyntheticEvent: event, string: name)
    renderItemContent: PropTypes.func // func(string: name, JsonSchema: schema, boolean: selected)
};

export default InspectorColumn;
