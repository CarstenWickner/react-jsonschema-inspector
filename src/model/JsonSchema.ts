import { isNonEmptyObject, deriveBaseUri } from "./utils";

import { RawJsonSchema, getValueFromRawJsonSchema } from "../types/RawJsonSchema";
import { ParserConfig } from "../types/ParserConfig";

const createRawSchemaFromBoolean = (value: boolean): RawJsonSchema => (value ? {} : { not: {} });

/**
 * Representation of a Json Schema, offering a number of convenient functions for traversing and extracting information.
 */
export class JsonSchema {
    /**
     * Raw JSON Schema.
     */
    readonly schema: RawJsonSchema;

    /**
     * Configuration steering how the json schema is being traversed/parsed.
     */
    readonly parserConfig: ParserConfig;

    /**
     * Collection of allowed references to other schemas/parts.
     */
    readonly scope: RefScope;

    /**
     * Constructor for a JsonSchema (wrapper).
     *
     * @param {boolean|RawJsonSchema} schema - the JSON Schema to represent
     * @param {ParserConfig} parserConfig - configuration affecting how the json schema is being traversed/parsed
     * @param {?RefScope} scope - collection of available $ref targets (will be generated based on `schema` if not provided)
     */
    constructor(schema: boolean | RawJsonSchema, parserConfig: ParserConfig, scope?: RefScope) {
        this.schema = typeof schema === "boolean" ? createRawSchemaFromBoolean(schema) : (schema as RawJsonSchema);
        this.parserConfig = parserConfig;
        this.scope = scope || new RefScope(this);
    }
}

/**
 * Helper class to facilitate looking-up re-usable sub-schemas via the "$ref" keyword. A single RefScope instance refers to one main schema
 * and its contained "$defs"/"definitions". Other main schemas maybe used by adding their respective RefScope instances via .addOtherRefScopes().
 */
export class RefScope {
    /**
     * Base URI for resolving references.
     */
    readonly baseUri: string | null;

    /**
     * Collection of available sub-schema to be referenced via "$ref" within the originating schema.
     */
    readonly internalRefs: Map<string, JsonSchema> = new Map();

    /**
     * Collection of available sub-schema to be referenced via "$ref" within the originating schema or from other schemas.
     */
    readonly externalRefs: Map<string, JsonSchema> = new Map();

    /**
     * Array of other scopes (e.g. in case of separate schemas being provided for that purpose).
     */
    otherScopes: Array<RefScope> = [];

    /**
     * Constructor collecting all available references to contained re-usable (sub-) schemas.
     *
     * @param {JsonSchema} schema - to collect $ref-erence-able sub-schemas from
     */
    constructor(schema?: JsonSchema) {
        this.baseUri = null;
        if (!schema || !isNonEmptyObject(schema.schema)) {
            return;
        }
        // can always self-reference via an empty fragment
        this.internalRefs.set("#", schema);
        // from JSON Schema Draft 6: "$id" replaces former "id"
        const mainAlias = getValueFromRawJsonSchema(schema.schema, "$id");
        let externalRefBase: string | null;
        if (mainAlias && !mainAlias.startsWith("#")) {
            // an absolute URI can be used both within the schema itself but also from other schemas
            const mainAliasWithFragment = mainAlias.endsWith("#") ? mainAlias : `${mainAlias}#`;
            const mainAliasWithoutFragment = mainAliasWithFragment.substring(0, mainAliasWithFragment.length - 1);
            this.externalRefs.set(mainAliasWithFragment, schema);
            this.externalRefs.set(mainAliasWithoutFragment, schema);
            // for definitions, there should always be the empty fragment between the URI and the definitions path
            externalRefBase = mainAliasWithFragment;
            this.baseUri = deriveBaseUri(mainAliasWithoutFragment);
        } else {
            // no valid alias provided
            externalRefBase = null;
        }
        // in draft 2019-09 the keyword "definitions" was renamed to "$defs"
        const definitionsKeyword = schema.schema.hasOwnProperty("$defs") ? "$defs" : "definitions";
        const definitions = getValueFromRawJsonSchema(schema.schema, definitionsKeyword);
        if (isNonEmptyObject(definitions)) {
            Object.keys(definitions).forEach((key) => {
                const definition = definitions[key];
                if (isNonEmptyObject(definition)) {
                    const subSchema: JsonSchema = new JsonSchema(definition, schema.parserConfig, this);
                    const subAlias = getValueFromRawJsonSchema(definition, "$id");
                    if (subAlias) {
                        // any alias provided within definitions will only be available as short-hand in this schema
                        this.internalRefs.set(subAlias, subSchema);
                    }
                    // from JSON Schema Draft 2019-09: "$anchor" for plain text references (that should no longer be provided via $id)
                    const anchor = getValueFromRawJsonSchema(definition, "$anchor");
                    if (anchor) {
                        // any alias provided within definitions will only be available as short-hand in this schema
                        this.internalRefs.set(`#${anchor}`, subSchema);
                    }
                    // can always reference schema in definitions by its path, starting from the empty fragment
                    this.internalRefs.set(`#/${definitionsKeyword}/${key}`, subSchema);
                    if (externalRefBase) {
                        // the convention was fulfilled and the top-level schema defined an absolute URI as its "$id"
                        // this allows referencing a schema in definitions by its path, starting from that URI
                        this.externalRefs.set(`${externalRefBase}/${definitionsKeyword}/${key}`, subSchema);
                        if (anchor) {
                            // from JSON Schema Draft 2019-09: "$anchor" can also be used in combination with base URI
                            this.externalRefs.set(`${externalRefBase}${anchor}`, subSchema);
                        }
                    }
                }
            });
        }
    }

    /**
     * Add other available scope.
     *
     * @param {RefScope} refScope - other reference scope that is available (at least its external $ref-erences)
     */
    addOtherScope(refScope: RefScope): void {
        this.otherScopes.push(refScope);
    }

    /**
     * Look-up a re-usable schema by its $ref-erence.
     *
     * @param {string} ref - the "$ref" value for which to look-up the associated (sub-)schema
     * @param {boolean} includeInternalRefs - whether the "$ref" value is from within the same main schema this RefScope belongs to
     * @returns {JsonSchema} the successfully looked-up reference (or null if no match was found)
     */
    findSchemaInThisScope(ref: string, includeInternalRefs = true): JsonSchema | undefined {
        return (includeInternalRefs && this.internalRefs.get(ref)) || this.externalRefs.get(ref);
    }

    /**
     * Look-up a re-usable schema by its $ref-erence.
     *
     * @param {string} ref - the "$ref" value for which to look-up the associated (sub-)schema
     * @returns {JsonSchema} the successfully looked-up reference
     * @throws error if no match was found
     */
    find(ref: string): JsonSchema {
        let result = this.findSchemaInThisScope(ref);
        const alternativeRef = !result && this.baseUri && !ref.startsWith("#") && `${this.baseUri}${ref}`;
        if (alternativeRef) {
            result = this.findSchemaInThisScope(alternativeRef);
        }
        if (!result) {
            this.otherScopes.some((otherScope) => {
                result = otherScope.findSchemaInThisScope(ref, false);
                if (!result && alternativeRef) {
                    result = otherScope.findSchemaInThisScope(alternativeRef, false);
                }
                return result;
            });
        }
        if (result) {
            return result;
        }
        throw new Error(`Cannot resolve $ref: "${ref}"${alternativeRef ? `/"${alternativeRef}"` : ""}`);
    }
}
