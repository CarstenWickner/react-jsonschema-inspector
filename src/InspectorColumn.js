import PropTypes from "prop-types";
import React from "react";
import classNames from "classnames";

import JsonSchema from "./JsonSchema";
import InspectorItem from "./InspectorItem";

const InspectorColumn = ({
    items, selectedItem, trailingSelection, filteredItems, onSelect, renderItemContent
}) => (
    <div
        className={classNames({
            "jsonschema-inspector-column": true,
            "with-selection": selectedItem,
            "trailing-selection": trailingSelection
        })}
        onClick={onSelect}
        role="presentation"
        tabIndex={-1}
    >
        {Object.keys(items).sort().map((name) => {
            const selected = name === selectedItem;
            const matchesFilter = filteredItems ? filteredItems.includes(name) : undefined;
            return (
                <InspectorItem
                    key={name}
                    name={name}
                    schema={items[name]}
                    selected={selected}
                    matchesFilter={matchesFilter}
                    onSelect={event => onSelect(event, name)}
                    renderContent={renderItemContent}
                />
            );
        })}
    </div>
);

InspectorColumn.propTypes = {
    items: PropTypes.objectOf(PropTypes.instanceOf(JsonSchema)).isRequired,
    selectedItem: ({ items, selectedItem }) => {
        if (selectedItem !== undefined && selectedItem !== null) {
            if (typeof selectedItem !== "string") {
                return new Error("`selectedItem` is not a `string`");
            }
            if (!items[selectedItem]) {
                return new Error("`selectedItem` is not part of `items`");
            }
        }
        // assume all ok
        return null;
    },
    trailingSelection: ({ selectedItem, trailingSelection }) => {
        if (trailingSelection && !selectedItem) {
            return new Error("`trailingSelection` is true while there is no `selectedItem`");
        }
        return null;
    },
    filteredItems: ({ items, filteredItems }) => {
        if (filteredItems !== undefined && filteredItems !== null) {
            if (!Array.isArray(filteredItems)) {
                return new Error("`filteredItems` is not an `array`");
            }
            if (filteredItems.some(singleItem => !items[singleItem])) {
                return new Error("`filteredItems` are not all part of `items`");
            }
        }
        return null;
    },
    onSelect: PropTypes.func.isRequired, // func(SyntheticEvent: event, string: name)
    renderItemContent: PropTypes.func // func({ string: name, boolean: hasNestedItems, boolean: selected, JsonSchema: schema })
};

InspectorColumn.defaultProps = {
    selectedItem: null,
    trailingSelection: false,
    filteredItems: null,
    renderItemContent: null
};

export default InspectorColumn;
