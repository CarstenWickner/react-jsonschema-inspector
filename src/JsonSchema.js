import RefScope from "./RefScope";
import {
    isNonEmptyObject, listValues, mapObjectValues, mergeObjects
} from "./utils";

class JsonSchema {
    /**
     * Alternative factory method to the constructor, returning "undefined" if the given schema is not a non-empty object.
     *
     * @param {Object} rawSchema the JSON Schema to represent
     * @param {Object} parserConfig configuration affecting how the JSON schema is being traversed/parsed
     * @param {RefScope} scope collection of available $ref targets
     * @returns {JsonSchema|undefined}
     */
    static createIfNotEmpty(rawSchema, parserConfig, scope) {
        return isNonEmptyObject(rawSchema) ? new JsonSchema(rawSchema, parserConfig, scope) : undefined;
    }

    /**
     * Raw JSON Schema
     */
    schema;

    /**
     * Configuration steering how the JSON schema is being traversed/parsed
     */
    parserConfig;

    /**
     * @type RefScope
     */
    scope;

    /**
     * Constructor for a JsonSchema (wrapper).
     *
     * @param {Object} schema the JSON Schema to represent
     * @param {Object} parserConfig configuration affecting how the JSON schema is being traversed/parsed
     * @param {?RefScope} scope collection of available $ref targets
     */
    constructor(schema, parserConfig, scope) {
        this.schema = schema;
        this.parserConfig = parserConfig;
        this.scope = scope || new RefScope(this);
    }

    getRelevantSchemaParts() {
        const { allOf, anyOf, oneOf } = this.schema;
        const includeAnyOf = anyOf && this.parserConfig && this.parserConfig.anyOf === "likeAllOf";
        const includeOneOf = oneOf && this.parserConfig && this.parserConfig.oneOf === "likeAllOf";
        if (allOf || includeAnyOf || includeOneOf) {
            let result = allOf;
            if (includeAnyOf) {
                result = result ? result.concat(anyOf) : anyOf;
            }
            if (includeOneOf) {
                result = result ? result.concat(oneOf) : oneOf;
            }
            return result;
        }
        return undefined;
    }

    /**
     * Look-up of values associated with a specific kind of field in the given JSON schema definition.
     * If the targeted schema is an array, the sub-schemas of its "items" will be ignored.
     *
     * @param {String} fieldName name/key of the field to look-up
     * @param {?Function} mergeFunction optional: how to combine two encountered values into one
     * @param {?Function} mappingFunction optional: converting the raw value before calling the mergeFunction
     * @return {*} result of the given mergeFunction over all encountered values of fields with the given name
     */
    getFieldValue(fieldName, mergeFunction = listValues, mappingFunction) {
        if (!isNonEmptyObject(this.schema)) {
            return undefined;
        }
        if (this.schema.$ref) {
            // schema variable is merely a reference to a separately defined schema
            const referencedSchema = this.scope.find(this.schema.$ref);
            // by convention, if $ref is specified, all other properties in the schema are being ignored
            return referencedSchema.getFieldValue(fieldName, mergeFunction, mappingFunction);
        }
        const rawValue = this.schema[fieldName];
        let value = mappingFunction ? mappingFunction(rawValue, this.parserConfig, this.scope) : rawValue;
        const schemaParts = this.getRelevantSchemaParts();
        if (schemaParts) {
            // schema variable is supposed to be combined with a given list of sub-schemas
            value = schemaParts
                .map(part => new JsonSchema(part, this.parserConfig, this.scope).getFieldValue(fieldName, mergeFunction, mappingFunction))
                .reduce(mergeFunction, value);
        }
        return value;
    }

    /**
     * Version of the getFieldValue() function to look-up of nested JSON schema definitions (while preserving the appropriate reference scope).
     *
     * This method is intended to be used to look-up fields like "items" or "additionalItems".
     *
     * @param {String} fieldName name/key of the field to look-up
     * @param {Function} mergeFunction optional: how to combine two encountered values into one
     * @return {*} result of the given mergeFunction over all encountered values of fields with the given name
     */
    getSchemaFieldValue(fieldName, mergeFunction = listValues) {
        return this.getFieldValue(fieldName, mergeFunction, JsonSchema.createIfNotEmpty);
    }

    /**
     * Determine whether this schema represents an array and if so, return the sub-schema describing its contents.
     *
     * @returns {JsonSchema} type of array contents or null if it is not an array or its content type could not be determined
     */
    getTypeOfArrayItems() {
        return this.getSchemaFieldValue("items") || this.getSchemaFieldValue("additionalItems") || null;
    }

    /**
     * Look-up of any nested sub-schemas that may contain information about "properties" in the given JSON schema definition.
     * If the given JSON schema definition refers is an array, it is assumed to not have relevant information.
     * Instead, the referenced schemas of its "Items" will be returned.
     *
     * This method is intended to be used to look-up fields like "properties" or "required" which are not associated with arrays.
     *
     * BEWARE: if a schema consists of multiple sub-schemas and only one of them defines an "items" field, the other sibling schemas
     * will still be included in the result list - they are just assumed to not have any relevant fields we are looking for.
     *
     * @returns {Array<JsonSchema>} array listing sub-schemas that may define properties about their children
     */
    getPropertyParentSchemas() {
        if (!isNonEmptyObject(this.schema)) {
            return [];
        }
        if (this.schema.$ref) {
            const referencedSchema = this.scope.find(this.schema.$ref);
            // this schema is just a reference to another separately defined schema
            return referencedSchema.getPropertyParentSchemas();
        }
        const schemaParts = this.getRelevantSchemaParts();
        if (schemaParts) {
            let returnedSchemas = [this];
            schemaParts.forEach((rawSchemaPart) => {
                returnedSchemas = returnedSchemas.concat(new JsonSchema(rawSchemaPart, this.parserConfig, this.scope).getPropertyParentSchemas());
            });
            return returnedSchemas;
        }
        if (isNonEmptyObject(this.schema.items)) {
            // unsupported: specifying array of schemas referring to entries at respective positions in described array
            // schema.items contains a single schema, that specifies the type of any value in the described array
            return new JsonSchema(this.schema.items, this.parserConfig, this.scope).getPropertyParentSchemas();
        }
        if (isNonEmptyObject(this.schema.additionalItems)) {
            // the given schema is an array which may specify its content type in "additionalItems"
            return new JsonSchema(this.schema.additionalItems, this.parserConfig, this.scope).getPropertyParentSchemas();
        }
        // if there are no nested schemas, only this schema itself may contain properties
        return [this];
    }

    /**
     * Extract the properties mentioned in this schema â€“ also considering all nested sub-schema by using getPropertyParentSchemas().
     *
     * @returns {Object.<String, JsonSchema>} collection of all properties mentioned in this schema
     */
    getProperties() {
        return this.getPropertyParentSchemas().map(({ schema, scope }) => {
            // properties is an Object.<String, raw-json-schema>
            const { properties, required = [] } = schema;
            const rawProperties = Object.assign(
                {},
                ...required.map(value => ({ [value]: true })),
                isNonEmptyObject(properties) ? properties : {}
            );
            return mapObjectValues(rawProperties, rawSchema => new JsonSchema(rawSchema, this.parserConfig, scope));
        }).reduce(mergeObjects, {});
    }
}

export default JsonSchema;
