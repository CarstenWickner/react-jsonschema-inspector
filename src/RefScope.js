import JsonSchema from "./JsonSchema";
import { isNonEmptyObject } from "./utils";

/**
 * Helper class to facilitate looking-up re-usable sub-schemas via the "$ref" keyword. A single RefScope instance refers to one main schema
 * and its contained "definitions". Other main schemas maybe used by adding their respective RefScope instances via .addOtherRefScopes().
 */
class RefScope {
    /**
     * Map.<String, JsonSchema>
     * collection of available sub-schema to be referenced via "$ref" within the originating schema.
     */
    internalRefs = new Map();

    /**
     * Map.<String, JsonSchema>
     * collection of available sub-schema to be referenced via "$ref" within the originating schema or from other schemas.
     */
    externalRefs = new Map();

    /**
     * Array.<RefScope>
     * Array of other scopes (e.g. in case of separate schemas being provided for that purpose).
     */
    otherScopes = [];

    /**
     * Constructor collecting all available references to contained re-usable (sub-) schemas.
     *
     * @param {JsonSchema} schema to collect $ref-erence-able sub-schemas from
     */
    constructor(schema) {
        if (!schema || !isNonEmptyObject(schema.schema)) {
            return;
        }
        // can always self-reference via an empty fragment
        this.internalRefs.set("#", schema);
        // from JSON Schema Draft 6: "$id" replaces former "id"
        const mainAlias = schema.schema.$id || schema.schema.id;
        let externalRefBase;
        if (mainAlias && !mainAlias.startsWith("#")) {
            // an absolute URI can be used both within the schema itself but also from other schemas
            const mainAliasWithFragment = mainAlias.endsWith("#") ? mainAlias : (`${mainAlias}#`);
            const mainAliasWithoutFragment = mainAliasWithFragment.substring(0, mainAliasWithFragment.length - 1);
            this.externalRefs.set(mainAliasWithFragment, schema);
            this.externalRefs.set(mainAliasWithoutFragment, schema);
            // for definitions, there should always by the empty fragment between the URI and the definitions path
            externalRefBase = mainAliasWithFragment;
        } else {
            // no valid alias provided
            externalRefBase = null;
        }
        const { definitions } = schema.schema;
        if (isNonEmptyObject(definitions)) {
            Object.keys(definitions).forEach((key) => {
                const definition = definitions[key];
                if (isNonEmptyObject(definition)) {
                    const subSchema = new JsonSchema(definition, schema.parserConfig, this);
                    // from JSON Schema Draft 6: "$id" replaces former "id"
                    const subAlias = definition.$id || definition.id;
                    if (subAlias) {
                        // any alias provided within "definitions" will only be available as short-hand in this schema
                        this.internalRefs.set(subAlias, subSchema);
                    }
                    // can always reference schema in definitions by its path, starting from the empty fragment
                    this.internalRefs.set(`#/definitions/${key}`, subSchema);
                    if (externalRefBase) {
                        // the convention was fulfilled and the top-level schema defined an absolute URI as its "$id"
                        // this allows referencing a schema in definitions by its path, starting from that URI
                        this.externalRefs.set(`${externalRefBase}/definitions/${key}`, subSchema);
                    }
                }
            });
        }
    }

    /**
     * Add other available scope.
     *
     * @param {RefScope} refScope other reference scope that is available (at least its external $ref-erences)
     */
    addOtherScope = (refScope) => {
        this.otherScopes.push(refScope);
    };

    /**
     * Add other available scopes.
     *
     * @param {Array.<RefScope>} refScopes other reference scopes that are available (at least their external $ref-erences)
     */
    addOtherScopes = (refScopes) => {
        refScopes.forEach(this.addOtherScope);
    };

    /**
     * Look-up a re-usable schema by its $ref-erence.
     *
     * @param {String} ref the "$ref" value for which to look-up the associated (sub-)schema
     * @param {Boolean} includeInternalRefs whether the "$ref" value is from within the same main schema this RefScope belongs to
     * @returns {JsonSchema} the successfully looked-up reference (or null if no match was found)
     */
    findSchemaInThisScope = (ref, includeInternalRefs = true) => (
        (includeInternalRefs && this.internalRefs.get(ref)) || this.externalRefs.get(ref)
    );

    /**
     * Look-up a re-usable schema by its $ref-erence.
     *
     * @param {String} ref the "$ref" value for which to look-up the associated (sub-)schema
     * @returns {JsonSchema} the successfully looked-up reference
     * @throws error if no match was found
     */
    find = (ref) => {
        let result = this.findSchemaInThisScope(ref);
        if (!result) {
            this.otherScopes.some((otherScope) => {
                result = otherScope.findSchemaInThisScope(ref, false);
                return result;
            });
        }
        if (result) {
            return result;
        }
        throw new Error(`Cannot resolve $ref: "${ref}"`);
    };
}

export default RefScope;
