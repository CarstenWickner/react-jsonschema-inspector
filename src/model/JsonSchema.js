import RefScope from "./RefScope";

/**
 * Representation of a Json Schema, offering a number of convenient functions for traversing and extracting information.
 */
export default class JsonSchema {
    /**
     * Raw JSON Schema.
     *
     * @type {object}
     */
    schema;

    /**
     * Configuration steering how the json schema is being traversed/parsed.
     *
     * @type {{oneOf: ?{groupTitle: ?string}, anyOf: ?{groupTitle: ?string}}}
     */
    parserConfig;

    /**
     * @type {RefScope}
     */
    scope;

    /**
     * Constructor for a JsonSchema (wrapper).
     *
     * @param {object} schema - the JSON Schema to represent
     * @param {object} parserConfig - configuration affecting how the json schema is being traversed/parsed
     * @param {?RefScope} scope - collection of available $ref targets (will be generated based on `schema` if not provided)
     */
    constructor(schema, parserConfig, scope) {
        this.schema = schema;
        this.parserConfig = parserConfig;
        this.scope = scope || new RefScope(this);
    }
}
