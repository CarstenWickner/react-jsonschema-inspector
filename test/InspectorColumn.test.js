import React from "react";
import { shallow } from "enzyme";
import RefScope from "../src/RefScope";
import InspectorColumn from "../src/InspectorColumn";
import JsonSchema from "../src/JsonSchema";

describe("renders correctly", () => {
    it("with minimal/default props", () => {
        const component = shallow(
            <InspectorColumn
                items={{
                    "Item One": new JsonSchema({
                        $id: "Schema One"
                    }, new RefScope({
                        definitions: { Target: {} }
                    })),
                    "Item Two": new JsonSchema({
                        $id: "Schema Two"
                    }, new RefScope({
                        definitions: { Target: {} }
                    }))
                }}
                onSelect={() => { }}
            />
        );
        expect(component).toMatchSnapshot();
    });
    it("with selection", () => {
        const component = shallow(
            <InspectorColumn
                items={{
                    "Item One": new JsonSchema(),
                    "Item Two": new JsonSchema()
                }}
                onSelect={() => { }}
                selectedItem="Item Two"
            />
        );
        expect(component.hasClass("with-selection")).toBe(true);
        expect(component.hasClass("trailing-selection")).toBe(false);
        expect(component.find("InspectorItem").at(0).prop("selected")).toBe(false);
        expect(component.find("InspectorItem").at(1).prop("selected")).toBe(true);
        expect(component.find("InspectorItem").at(0).prop("matchesFilter")).toBeUndefined();
        expect(component.find("InspectorItem").at(1).prop("matchesFilter")).toBeUndefined();
    });
    it("with trailing selection", () => {
        const component = shallow(
            <InspectorColumn
                items={{
                    "Item One": new JsonSchema(),
                    "Item Two": new JsonSchema()
                }}
                onSelect={() => { }}
                selectedItem="Item Two"
                trailingSelection
            />
        );
        expect(component.hasClass("with-selection")).toBe(true);
        expect(component.hasClass("trailing-selection")).toBe(true);
        expect(component.find("InspectorItem").at(0).prop("selected")).toBe(false);
        expect(component.find("InspectorItem").at(1).prop("selected")).toBe(true);
        expect(component.find("InspectorItem").at(0).prop("matchesFilter")).toBeUndefined();
        expect(component.find("InspectorItem").at(1).prop("matchesFilter")).toBeUndefined();
    });
    it("with filter", () => {
        const component = shallow(
            <InspectorColumn
                items={{
                    "Item One": new JsonSchema(),
                    "Item Two": new JsonSchema()
                }}
                onSelect={() => { }}
                filteredItems={["Item Two"]}
            />
        );
        expect(component.hasClass("with-selection")).toBe(false);
        expect(component.hasClass("trailing-selection")).toBe(false);
        expect(component.find("InspectorItem").at(0).prop("matchesFilter")).toBe(false);
        expect(component.find("InspectorItem").at(1).prop("matchesFilter")).toBe(true);
    });
});
describe("calls onSelect", () => {
    let onSelect;
    let component;
    beforeEach(() => {
        onSelect = jest.fn(() => { });
        component = shallow(
            <InspectorColumn
                items={{
                    "Item One": new JsonSchema()
                }}
                onSelect={onSelect}
            />
        );
    });

    it("clearing selection when clicking on column", () => {
        component.find(".jsonschema-inspector-column").prop("onClick")({});
        expect(onSelect.mock.calls).toHaveLength(1);
        // expect no second parameter indicating selected item
        expect(onSelect.mock.calls[0]).toHaveLength(1);
    });
    it("setting selection when clicking on item", () => {
        component.find("InspectorItem").prop("onSelect")({});
        expect(onSelect.mock.calls).toHaveLength(1);
        // expect no second parameter indicating selected item
        expect(onSelect.mock.calls[0]).toHaveLength(2);
        expect(onSelect.mock.calls[0][1]).toEqual("Item One");
    });
});
describe("failing PropType validation", () => {
    it("for invalid selectedItem", () => {
        const component = () => (
            <InspectorColumn
                items={{
                    "Item One": new JsonSchema()
                }}
                onSelect={() => { }}
                selectedItem={0}
            />
        );
        expect(component).toThrowError("Warning: Failed prop type: `selectedItem` is not a `string`\n"
            + "    in InspectorColumn");
    });
    it("for selectedItem not being part of items", () => {
        const component = () => (
            <InspectorColumn
                items={{
                    "Item One": new JsonSchema()
                }}
                onSelect={() => { }}
                selectedItem="Item Two"
            />
        );
        expect(component).toThrowError("Warning: Failed prop type: `selectedItem` is not part of `items`\n"
            + "    in InspectorColumn");
    });
    it("for trailingSelection without selectedItem", () => {
        const component = () => (
            <InspectorColumn
                items={{
                    "Item One": new JsonSchema()
                }}
                onSelect={() => { }}
                trailingSelection
            />
        );
        expect(component).toThrowError("Warning: Failed prop type: `trailingSelection` is true while there is no `selectedItem`\n"
            + "    in InspectorColumn");
    });
    it("for invalid filteredItems", () => {
        const component = () => (
            <InspectorColumn
                items={{
                    "Item One": new JsonSchema()
                }}
                onSelect={() => { }}
                filteredItems={1}
            />
        );
        expect(component).toThrowError("Warning: Failed prop type: `filteredItems` is not an `array`\n"
            + "    in InspectorColumn");
    });
    it("for filteredItems not all being part of items", () => {
        const component = () => (
            <InspectorColumn
                items={{
                    "Item One": new JsonSchema()
                }}
                onSelect={() => { }}
                filteredItems={["Item One", "Item Two"]}
            />
        );
        expect(component).toThrowError("Warning: Failed prop type: `filteredItems` are not all part of `items`\n"
            + "    in InspectorColumn");
    });
});
