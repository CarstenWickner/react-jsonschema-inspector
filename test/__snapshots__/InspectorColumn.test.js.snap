// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly with minimal/default props 1`] = `
<div
  className="jsonschema-inspector-column"
  onClick={[Function]}
  role="presentation"
  tabIndex={-1}
>
  <InspectorItem
    key="Item One"
    name="Item One"
    onSelect={[Function]}
    renderContent={null}
    schema={
      JsonSchema {
        "parserConfig": Object {},
        "schema": Object {
          "$id": "Schema One",
        },
        "scope": RefScope {
          "addOtherScope": [Function],
          "externalRefs": Map {},
          "internalRefs": Map {},
          "otherScopes": Array [],
        },
      }
    }
    selected={false}
  />
  <InspectorItem
    key="Item Two"
    name="Item Two"
    onSelect={[Function]}
    renderContent={null}
    schema={
      JsonSchema {
        "parserConfig": Object {},
        "schema": Object {
          "$id": "Schema Two",
        },
        "scope": RefScope {
          "addOtherScope": [Function],
          "externalRefs": Map {},
          "internalRefs": Map {},
          "otherScopes": Array [],
        },
      }
    }
    selected={false}
  />
</div>
`;
